# Import Libraries & Data

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# establish path to project folder
path = r'C:\Users\BrianDavis\OneDrive - abetterchicago.org\Documents\Data Analysis\Achievement 5 (Python)\04-2023 Instacart Basket Analysis'

In [3]:
# import "orders_products_grouped" pkl file
df = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_grouped.pkl'))

# Manipulate Data (Assignment 5.8)

### step 2 - In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [4]:
#group dataframe by department_id and calculate mean of order number
df.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### step 3 - Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

#### Observations: the results from this aggregation on the entire dataset differ from the aggregation done on the subset: notably, department 17 (household) has a much higher average (16.8) in the full dataset vs only 12.6 in the subset. All other averages changed, but this is to be expected given the increase in number of rows being analyzed

### step 4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions. 

### step 5 - The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [5]:
#calculate statistics around product prices for each loyalty category
df.groupby('loyalty_flag').agg({'prices':['mean']})

prices
                      mean
loyalty_flag              
Loyal customer    7.772831
New customer      7.800320
Regular customer  7.797431

### step 6 - The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders.

In [6]:
#using transform(), add a new column with the average prices paid for each user
df['average_price'] = df.groupby('user_id')['prices'].transform(np.mean)

In [7]:
#using loc(), create a spending flag based on average prices paid for each user

#establish conditions
df.loc[df['average_price'] >= 10, 'spending flag'] = 'High spender'

In [8]:
#establish conditions
df.loc[df['average_price'] < 10, 'spending flag'] = 'Low spender'

In [9]:
#call relevant columns to check 
df[['user_id', 'average_price', 'spending flag']].head(50)

user_id  average_price spending flag
0         1       6.367797   Low spender
1         1       6.367797   Low spender
2         1       6.367797   Low spender
3         1       6.367797   Low spender
4         1       6.367797   Low spender
5         1       6.367797   Low spender
6         1       6.367797   Low spender
7         1       6.367797   Low spender
8         1       6.367797   Low spender
9         1       6.367797   Low spender
10       15       3.980556   Low spender
11       15       3.980556   Low spender
12       15       3.980556   Low spender
13       15       3.980556   Low spender
14       15       3.980556   Low spender
15       19       8.372059   Low spender
16       19       8.372059   Low spender
17       19       8.372059   Low spender
18       21       7.012683   Low spender
19       31       8.137793   Low spender
20       31       8.137793   Low spender
21       43       7.444898   Low spender
22       43       7.444898   Low spender
23       52       6.356287   Low spender
24       52       6.356287   Low spender
25       52       6.356287   Low spender
26       52       6.356287   Low spender
27       52       6.356287   Low spender
28       52       6.356287   Low spender
29       52       6.356287   Low spender
30       52       6.356287   Low spender
31       52       6.356287   Low spender
32       52       6.356287   Low spender
33       52       6.356287   Low spender
34       52       6.356287   Low spender
35       52       6.356287   Low spender
36       52       6.356287   Low spender
37       67       7.656790   Low spender
38       67       7.656790   Low spender
39       67       7.656790   Low spender
40       67       7.656790   Low spender
41       67       7.656790   Low spender
42       67       7.656790   Low spender
43       67       7.656790   Low spender
44       67       7.656790   Low spender
45       67       7.656790   Low spender
46       67       7.656790   Low spender
47       67       7.656790   Low spender
48       67       7.656790   Low spender
49       67       7.656790   Low spender

In [10]:
# count values of relevant columns to check
df['spending flag'].value_counts(dropna = False)

Low spender     32285165
High spender      119694
Name: spending flag, dtype: int64

### step 7 - In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 

In [11]:
#using transform(), add a new column with the median order frequency of each user 
df['order_frequency'] = df.groupby('user_id')['days_since_prior_order'].transform(np.median)

In [12]:
#using loc(), create an order frequency flag based on median days between orders for each user

#establish conditions
df.loc[df['order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [13]:
#establish conditions
df.loc[(df['order_frequency'] <= 20) & (df['order_frequency'] > 10), 'order_frequency_flag'] = 'Regular customer'

In [14]:
#establish conditions
df.loc[df['order_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [15]:
#call relevant columns to check
df[['user_id', 'order_frequency', 'order_frequency_flag']].head(50)

user_id  order_frequency   order_frequency_flag
0         1             20.5  Non-frequent customer
1         1             20.5  Non-frequent customer
2         1             20.5  Non-frequent customer
3         1             20.5  Non-frequent customer
4         1             20.5  Non-frequent customer
5         1             20.5  Non-frequent customer
6         1             20.5  Non-frequent customer
7         1             20.5  Non-frequent customer
8         1             20.5  Non-frequent customer
9         1             20.5  Non-frequent customer
10       15             10.0      Frequent customer
11       15             10.0      Frequent customer
12       15             10.0      Frequent customer
13       15             10.0      Frequent customer
14       15             10.0      Frequent customer
15       19              6.0      Frequent customer
16       19              6.0      Frequent customer
17       19              6.0      Frequent customer
18       21              7.5      Frequent customer
19       31              6.0      Frequent customer
20       31              6.0      Frequent customer
21       43             10.0      Frequent customer
22       43             10.0      Frequent customer
23       52              8.0      Frequent customer
24       52              8.0      Frequent customer
25       52              8.0      Frequent customer
26       52              8.0      Frequent customer
27       52              8.0      Frequent customer
28       52              8.0      Frequent customer
29       52              8.0      Frequent customer
30       52              8.0      Frequent customer
31       52              8.0      Frequent customer
32       52              8.0      Frequent customer
33       52              8.0      Frequent customer
34       52              8.0      Frequent customer
35       52              8.0      Frequent customer
36       52              8.0      Frequent customer
37       67              7.0      Frequent customer
38       67              7.0      Frequent customer
39       67              7.0      Frequent customer
40       67              7.0      Frequent customer
41       67              7.0      Frequent customer
42       67              7.0      Frequent customer
43       67              7.0      Frequent customer
44       67              7.0      Frequent customer
45       67              7.0      Frequent customer
46       67              7.0      Frequent customer
47       67              7.0      Frequent customer
48       67              7.0      Frequent customer
49       67              7.0      Frequent customer

In [16]:
# count values of relevant columns to check
df['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [17]:
df['order_day_of_week'].value_counts()

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

#### Observation: Saturday was the most common day of the week, followed by Sunday. Wednesday was the least common day of the week. 

# Export Data

In [18]:
df.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_grouped_2.pkl'))